core/src/main/java/io/opencensus/tags/Tag.java

```diff
@@ -0,0 +1,202 @@
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.tags;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.Function;
+import io.opencensus.tags.TagKey.TagKeyBoolean;
+import io.opencensus.tags.TagKey.TagKeyLong;
+import io.opencensus.tags.TagKey.TagKeyString;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/** {@link TagKey} paired with an optional value. */
+@Immutable
+public abstract class Tag {
+
+  /**
+   * Returns the tag's key.
+   *
+   * @return the tag's key.
+   */
+  public abstract TagKey getKey();
+
+  /**
+   * Applies a function to the tag's key and value. The function that is called depends on the type
+   * of the tag. This is similar to the visitor pattern.
+   *
+   * @param stringFunction the function to call when the tag has a {@code String} value.
+   * @param longFunction the function to call when the tag has a {@code Long} value.
+   * @param booleanFunction the function to call when the tag has a {@code Boolean} value.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag's type.
+   */
+  public <T> T match(
+      Function<? super TagString, T> stringFunction,
+      Function<? super TagLong, T> longFunction,
+      Function<? super TagBoolean, T> booleanFunction) {
+    return matchWithDefault(
+        stringFunction, longFunction, booleanFunction, TagUtils.<T>throwAssertionError());
+  }
+
+  /**
+   * Applies a function to the tag's key and value. This method is like {@link #match(Function,
+   * Function, Function)}, except that it has a default case, for backwards compatibility when tag
+   * types are added.
+   *
+   * @param stringFunction the function to call when the tag has a {@code String} value.
+   * @param longFunction the function to call when the tag has a {@code Long} value.
+   * @param booleanFunction the function to call when the tag has a {@code Boolean} value.
+   * @param defaultFunction the function to call when the tag has a value other than {@code String},
+   *     {@code Long}, or {@code Boolean}.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag's type.
+   */
+  // TODO(sebright): How should we deal with the possibility of adding more tag types in the future?
+  //                 Will this method work? What type of parameter should we use for the default
+  //                 case? Should we remove the non-backwards compatible "match" method?
+  public abstract <T> T matchWithDefault(
```

Comment:
Can you copy the discussion about the matchWithDefault() to the 'Tagged Unions in Java' design doc? I'd like to capture it there too and it would make it easier to discuss.

One of the advantages of just having the match() method (w/o default) is that when extending the type (which shouldn't happen often) you will get compile time errors for any code that you haven't updated to handle the new cases instead of potentially having them handled by the matchWithDefault() method - I think this is actually preferable.